src/pallene/pallenelib.lua

```diff
@@ -90,12 +154,267 @@ static lua_Number  pallene_math_log(lua_Integer x, lua_Integer base);
+            if(lua_rawequal(L, fn_idx, -1)) {
+                /* Remove value and keep name. */
+                lua_pop(L, 1);
+
+                return true;
+            }
+            /* If not go one level deeper and get the value recursively. */
+            if(_findfield(L, fn_idx, level - 1)) {
+                /* Remove the table but keep name. */
+                lua_remove(L, -2);
+
+                /* Add a "." in between. */
+                lua_pushliteral(L, ".");
+                lua_insert(L, -2);
+
+                /* Concatenate last 3 values, resulting "table.some_func". */
+                lua_concat(L, 3);
+
+                return true;
+            }
+        }
+
+        /* Pop the value. */
+        lua_pop(L, 1);
+    }
+
+    return false;
+}
+
+/* Pushes a function name if found in the global table and returns true.
+   Returns false otherwise. */
+/* Expects the funtion to be pushed in the stack. */
+static bool _pgf_name(lua_State *L) {
+    int top = lua_gettop(L);
+
+    lua_pushglobaltable(L);
+
+    if(_findfield(L, top, 2)) {
+        lua_remove(L, -2);
+
+        return true;
+    }
+
+    lua_pop(L, 1);
+    return false;
+}
+
+/* Private routines end. */
+
+static void pallene_tracer_frameenter(lua_State *L, pt_frame_t *restrict frame) {
+    /* Retrieve the end of the stack. */
+    lua_getglobal(L, "__pallene_tracer_stack");
+    pt_frame_t *stack = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 1);
+
+    /* If there is no frame in the stack. */
+    /* No matter what type of frame we got (Lua or plain C), it will be
+       in the general stack. */
+    if(l_unlikely(stack == NULL)) {
+        frame->prev = NULL;
+        stack = frame;
+
+        goto out;
+    }
+
+    frame->prev = stack;
+    stack = frame;
+
+out:
+    lua_pushlightuserdata(L, stack);
+    lua_setglobal(L, "__pallene_tracer_stack");
+}
+
+static void pallene_tracer_global_setline(lua_State *L, int line) {
+    lua_getglobal(L, "__pallene_tracer_stack");
+    pt_frame_t *stack = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 1);
+
+    if(stack != NULL)
+        stack->line = line;
+}
+
+static void pallene_tracer_setline(pt_frame_t *restrict frame, int line) {
+    frame->line = line;
+}
+
+static void pallene_tracer_frameexit(lua_State *L) {
+    /* Retrieve the end of the stack. */
+    lua_getglobal(L, "__pallene_tracer_stack");
+    pt_frame_t *stack = (pt_frame_t *) lua_topointer(L, -1);
+    lua_pop(L, 1);
+
+    /* We are popping the very last frame. */
+    if(stack->prev == NULL) {
+        stack = NULL;
+        goto out;
+    }
+
+    stack = stack->prev;
+
+out:
+    lua_pushlightuserdata(L, stack);
+    lua_setglobal(L, "__pallene_tracer_stack");
+}
+
+static int pallene_tracer_debug_traceback(lua_State *L) {
+    const char *message = lua_tostring(L, 1);
+    fprintf(stderr, "Runtime error: %s\nStack traceback: \n", message);
+
+    /* Lua: 1, Pallene: 0 */
+    int context = 1;
```

Comment:
I wonder if we can structure our traversal as two nested loops and get rid of the context flag and some of the gotos.

    for each frame in the Lua stack
        if this frame is linked to the Pallene stack
            for each frame in the Pallene stack block
                 print C frame
        else
            print Lua frame

Reply:
Doable.